### ISSUE TITLE
Make scipy a required dependency

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#12448]
I would like to suggest making scipy a required dependency of the core package.

There are a number of sub-packages in astropy where we use scipy where it's not obvious that would be the case and where users will encounter unexpected ImportErrors if it is not installed. For instance the following code snippets from the docs require scipy:

**astropy.coordinates**

```python
>>> c = SkyCoord(ra=ra1*u.degree, dec=dec1*u.degree, distance=distance1*u.kpc)
>>> catalog = SkyCoord(ra=ra2*u.degree, dec=dec2*u.degree, distance=distance2*u.kpc)
>>> idx, d2d, d3d = c.match_to_catalog_3d(catalog)  # REQUIRES SCIPY
```

**astropy.cosmology**

```python
>>> from astropy.cosmology import WMAP9 as cosmo
>>> cosmo.H(0)
<Quantity 69.32 km / (Mpc s)>
>>> cosmo.kpc_proper_per_arcmin(3)  # REQUIRES SCIPY
<Quantity 472.97709620405266 kpc / arcmin>
```

**astropy.units**
```python
>>> from astropy.cosmology import WMAP9
>>> z = 1100 * cu.redshift
>>> z.to(u.K, cu.with_redshift(WMAP9))
<Quantity 3000.225 K>
>>> z.to(u.Mpc, cu.with_redshift(WMAP9, distance="luminosity"))   # REQUIRES SCIPY
<Quantity 15418438.76317008 Mpc>
```

**astropy.convolution**
```python
>>> gauss_oversample = Gaussian1DKernel(3, mode='oversample', factor=10)
>>> gauss_integrate = Gaussian1DKernel(3, mode='integrate')   # REQUIRES SCIPY
```

**astropy.modeling**

scipy is also used extensively in modeling!

In all of these examples, there is no way the user would know in advance that scipy is required, and they will stumble on these issues and have to install scipy to get them to work. Now that scipy is needed for some of the functionality in at least five sub-packages, I think it is time to consider requiring it. In total, ``from scipy.`` appears 76 times in the astropy core package, and ``doctest-requires:: scipy`` appears 36 times!

As packaging has improved tremendously since the start of the project, and wheels and conda packages are available for scipy for many platforms, I think it would be reasonable to start requiring scipy and would result in fewer surprise errors for users.

The philosophy here is the same as why we added PyYAML as a required dependency - in many cases users couldn't have suspected this dependency.

To give a counter example of something I think doesn't need to be a required dependency, astropy.visualization provides lots of functionality to expand matplotlib, but in many cases matplotlib needs to be installed anyway for the user to want to use the functionality. Take a look at the following example:

```
    from astropy import units as u
    from astropy.visualization import quantity_support
    quantity_support()
    from matplotlib import pyplot as plt
    plt.figure(figsize=(5,3))
    plt.plot([1, 2, 3] * u.m)
```

Here matplotlib is imported explicitly anyway, and a user wouldn't be trying to use ``quantity_support`` without it installed.

Another counter-example would be the table reader/writer for HDF5 - in this case it's ok to error if h5py is not installed with an explicit error that says 'h5py is required to read/write HDF5' files.

Perhaps this requires an APE or a policy document describing some basic rules/tests we can use to determine if a dependency should be required.

**Comment by nstarman** [comment]
It would be very convenient if scipy were a runtime dependency.
From what I understand, the major reason why it hasn't already been made one is the issue of installation on various systems. Perhaps it could be tabulated what system astropy runs on versus where scipy can be installed...

**Comment by pllim** [comment]
I was neutral but now I am 👎 on this. I have a package that has scipy as required dependency and now the release wheel building workflow is broken on `ubuntu-latest` because it is somehow trying to build `scipy` from source. I am still investigating, but this gives you an idea of the possible headaches we have to deal with.

**Comment by pllim** [comment]
Update: Okay, I traced the scipy build problem to `cibuildwheels` automatically trying to build for `musllinux`. Disabling that as follows fixed the problem. So I am neutral again.

```yaml
CIBW_SKIP: '*-musllinux_*'
```

**Comment by adrn** [comment]
I think the case for making it an `install_requires` dependency is convincing enough, but I don't work on systems where scipy is difficult to install. Before we presuppose pushback, maybe it's worth sending a question to astropy-dev to see if  people in those situations anticipate real issues?

**Comment by pllim** [comment]
Posted to astropy-dev: https://groups.google.com/g/astropy-dev/c/T5gskNMJpx8

**Comment by mhvk** [comment]
Scipy adds 85 MB to the installation while astropy itself is 19 MB and numpy is 28 MB (well, the debian packages are). That does feel like the tail wagging the dog! So, I'm not too keen.

By the way, `units` does not require `scipy` - the example above uses cosmology... But at some level that proves the point - it is not obvious from the example that scipy would be needed! But anyway, I'd say that `cosmology`, `convolution` and `modeling` are the ones where `scipy` is truly needed.

Note that only a little while ago, I added a "recommended" category in `setup.cfg`, and updated our [installation instructions](https://docs.astropy.org/en/latest/install.html) to suggest one do
```
pip install astropy[recommended]
```

Unfortunately, I don't know if there is a similarly easy way to get a more minimal installation if we were to add scipy to the installation requirements.

**Comment by dstansby** [comment]
One possible approach is to have a set of 'core' dependencies, and then a set of required dependencies per sub-module. So e.g. `pip install astropy[cosmology]` or `pip install astropy[convolution]` would install `scipy`, but `pip install astropy` wouldn't.

**Comment by mhvk** [comment]
@dstansby - that sounds dangerously like something we have at times considered, which is to package astropy in parts that somewhat respect the internal hierarchy, i.e., `pip install astropy-cosmology` which pulls in `astropy-core` as well as `scipy` (and where `astropy-core` in turn may depend on `astropy-units`, as units & constants basically do not depend on anything else in astropy). But obviously very unclear if this is worth the likely considerable effort!

**Comment by dhomeier** [comment]
> Scipy adds 85 MB to the installation while astropy itself is 19 MB and numpy is 28 MB (well, the debian packages are). That does feel like the tail wagging the dog! So, I'm not too keen.

That was one of my first 3 thoughts on this issue as well. Installed size I get on macOS is 70 / 35 / 31 MB, so that's not quite as bad.
The other 2 were, to me it seems so weird not using any Scipy functionality it may as well be always installed;
then I felt it would expand the dependency chain quite a bit beyond our original Numpy-only situation.
But actually there are no other Python runtime dependencies, and the only potential heavy-weight requirements are probably a BLAS, which is already required by Numpy, and Suitesparse.
OTOH there are certainly still many users out there who are using Astropy chiefly for `io.fits` or similar, so for them doubling effective install size of an already relatively large package is not negligible.
At least we may probably trust that there is some awareness of install size issue on the Scipy side as well – scipy/PROPACK#2.

**Comment by mhvk** [comment]
Debian doesn't bundle our compiled libraries (like `libcfitsio` and `libwcs`), which may explain part of the difference. It is good that scipy is thinking about how to reduce its footprint. Although it doesn't seem test files at least in Debian, but rather just cython compiled libraries... Just to be sure, I checked and at least the Debian installation of astropy is pretty good: the largest file by far is the IERS table.

Anyway, this is getting a bit off-topic. Apologies!

**Comment by dhomeier** [comment]
> @dstansby - that sounds dangerously like something we have at times considered, which is to package astropy in parts that somewhat respect the internal hierarchy,

I rather understood the suggestion as labels for various bundles, i.e. `astropy[cosmology]`, `astropy[convolution]` or `astropy[modeling]` would just include the additional dependencies for the respective sub-packages, so `astropy[recommended]` would become something like a superset of them all.
Splitting the core package itself into partial installations also sounds like a dangerous can of worms to me.

**Comment by hamogu** [comment]
I think adding scipy as a hard dependency is OK at this point. I have not seen a system in years where an extra < 100 MB install would be a problem. I install scipy into my environments for almost any project anyway, because sooner or later, I always need the interpolation or integration or one of the special functions from scipy. 

I think I'm a fairly typical user (laptop or linux dektop with sufficient disk space, using conda environments, no HPC or exotic systems) and to me it seems that in @astrofrog list only two options are worthwhile:
- keeps a is now
- make scipy a full dependency

Anything in between strikes me a a lot of extra effort and mental overhead especially for casual developers (scipy is required, but we still need to make sure tests pass without?) for limited gain.

Note: `astropy.stats.poisson_conf_interval(... interval=‘kraft-burrows-nousek’)` has a dependency of either scipy or mpmath, where mpmath is actually preferred because of the higher precision you can get. So, that is an example where "rewrite to use a light-weight library instead of scipy" is in place already; but unfortunately, mpmath won't do all the things we need scipy for.

**Comment by hamogu** [comment]
I think it would also be good to discuss this in a dev telecon. I'll put it on the agenda for next time.

**Comment by mhvk** [comment]
Good to discuss in telecon. As I'll be teaching and less likely to be available (though I'll try), my 2¢ is that I'd like to see some actual user confusion about import errors before changing. As I noted above, there is a recommended and documented procedure to pull in packages we strongly recommend users have, one that has been around only for a little while. I think it makes sense to see whether with those instructions now present, there are actual issues before changing.

**Comment by hamogu** [comment]
I typically install through conda. The scipy package I got there is about 20 MB. Up till right now, I did not know that we have a list of optional dependencies in the installation instruction; I guess I should be reading those more carefully more often. Personally, I prefer a few more required dependencies that I get automatically over more or less surprising failures because I only have the minimal dependencies installed.

**Comment by dhomeier** [comment]
A bit different from what I get on macOS:
> 48100	opt/miniconda3/envs/aptest/lib/python3.9/site-packages/astropy/
27292	opt/miniconda3/envs/aptest/lib/python3.9/site-packages/numpy/
70200	opt/miniconda3/envs/aptest/lib/python3.9/site-packages/scipy/
  145592 total

So, while still not dramatic, scipy is doubling the install size on the python package side.
OTOH the current conda on x64 installs just for a minimal numpy installation nearly 1 GB(!) of mkl library versions, probably ¾ of them for architectures my hardware either does not need or cannot use (for multiple envs they seem to be kept as hardlinks, so the impact on disk space is not quite as outrageous). As long as conda policies are like this, scipy is probably the least to worry about. But it would still matter for folks installing with pip, or other package managers.

**Comment by astrofrog** [comment]
> Unfortunately, I don't know if there is a similarly easy way to get a more minimal installation if we were to add scipy to the installation requirements.

```
pip install astropy --no-deps
```

😄

**Comment by mhvk** [comment]
Reading the very long thread on the possibility of "default extras" at https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/113, I cannot help but notice the suggestion that of course one should just mimic debian...

> If the idea is that a package has some dependencies that aren’t strictly necessary but will be wanted in most configurations, then something like Debian’s “Recommends:” field would make sense (i.e., a new variant of install_requires rather than a new variant of extras_require). The semantics would be that normally all install_recommends packages will be installed, unless you run pip --no-recommended or something like that.

**Comment by mhvk** [comment]
But more seriously, if the decision is to include `scipy`, I can of course live with that - all we'd need in that case is in the docs to have a working substitute for the no-scipy case (along the lines of what you have but that includes `pyerfa`, etc.)

**Comment by hamogu** [comment]
For reference: @astrofrog was inspired by a twitter thread where a real user (https://twitter.com/mileslucas_/status/1436452596982554624) reported the optional scipy dependency as a problem.

**Comment by robertdstein** [comment]
Can I just comment that I am a user building a python package in a clean environment that requires astropy, who was somewhat perplexed to find a Scipy import error. I did not have scipy installed, though obviously it is trivial to do so, but now I have to explicitly add scipy as a dependency in my project (and manage versions etc in CI) to fix what I see as an internal astropy problem. I only found this thread because I wanted to open an issue on this myself. As one random user, I would strongly favour that, as scipy is required for many functions, it should be listed as a dependency and installed automatically when I do `pip install astropy`. I had sort of just assumed that this happened already.

**Comment by hamogu** [comment]
@robertdstein Thanks for your feedback. That is very valuable to know. One of the problems in (astropy) development is that many users do not tell us about their expectations and many of our conversations start with "A typical user might ...", which is not a very good basis to make the call what dependency should be required vs. optional.

**Comment by mhvk** [comment]
Indeed, thanks for the feedback! So, the current recommended procedure is to depend on `astropy[recommended]` if your package needs also parts that depend on scipy, etc.

I just wish it was possible to make `scipy` default and have an `astropy[minimal]`!

**Comment by karllark** [comment]
I am surprised to learn that scipy is not required for astropy.  Like others, I just assumed it was.  Hence, I'm in favor of option 1).

**Comment by astrofrog** [comment]
So coming back to this, I'd like to suggest following options 3 and 4 - namely that we list scipy as a required dependency but only on common platforms where wheels are available which I think would satisfy most regular users of astropy while not forcing people on more exotic platforms to build scipy from source. We then would make sure that the test suite can run without scipy installed. Finally, we would add a comment in the requirements that we recommend that distribution maintainers, e.g. conda, debian etc include scipy as a required or default dependency).

**Comment by aragilar** [comment]
@astrofrog What platforms are you defining as exotic?

Doing 5 for all optional dependencies would seem like a good idea anyway (independent of whether scipy becomes required or not), but I'm not sure that actually answers https://twitter.com/mileslucas_/status/1436452596982554624, as it seems people don't realise how many (optional) dependencies astropy actually has.

I wonder if the long term thing to do is:
a) be significantly more strict around what gets into the astropy core package
b) look at what can be moved to/become a coordinated package (e.g. does cosmology need to be in the core package, should it rather be a coordinated package, maybe look at modelling also)
c) split out API docs from usage/tutorial docs, and use coordinated packages throughout (noting explicitly that they explicitly need to be installed)
d) create an astropy-all (astropy-full? astropy-ecosystem?) metapackage which installs the core astropy package as well as all the coordinates ones. This can then be used for interactive work to get people started in the ecosystem, allowing the core to be slimmed down (e.g. for use in containers)

**Comment by astrofrog** [comment]
@aragilar - what I really mean is that we could include scipy as a dependency for platforms where wheels are usually available.

**Comment by astrofrog** [comment]
I wonder if this actually warrants an APE with some general best practices/guidelines we want to follow as it would be good to be consistent across all the different astropy packages. I'll try and draft up something, but this is clearly not something we are going to decide ahead of 5.1.

**Comment by astrofrog** [comment]
On a related note, https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/127 is moving towards a PEP - based on the discussion it looks like one could define e.g. a ``recommended`` extras and then allow users to do a minimal install by doing ``pip install astropy[-recommended]``.

**Comment by mhvk** [comment]
That would be a perfect solution! But even so probably good to have some process in place for how we add to that "default extras" list (and decide between that and required dependencies).

**Comment by dhomeier** [comment]
Bringing some activity here again, assuming we do set a timeline for requiring SciPy (realising this may never happen if we decide to follow the `recommended` model), how should planned changes that will require SciPy warn about this, e.g.
https://github.com/astropy/astropy/pull/14013#discussion_r1146178233 ?

**Comment by astrofrog** [comment]
I am working on an APE, more soon!

**Comment by nstarman** [comment]
Is this intended for 6.0? It would enable a lot of simplifications for cosmology!

**Comment by pllim** [comment]
@nstarman , I haven't seen any APE yet, so... not sure.

**Comment by dhomeier** [comment]
The APE would be about defining a `recommended` extra package, right?

**Comment by hamogu** [comment]
Do we need an APE for this? I don't think we did last time we added dependencies. That's not to say we can't write one, just that it might not reach the level where it's *required* to have an APE.

**Comment by nstarman** [comment]
I can open a PR right now to make scipy required...

**Comment by dhomeier** [comment]
There has been no consensus on making it a required dependency, so in https://github.com/astropy/astropy/issues/12448#issuecomment-1029156776 and following the discussion evolved towards ensuring Scipy is installed by default on platforms where it is easily available, while still allowing platforms where it isn't to get a basic working Astropy installation. So this is somewhat more complex than just adding another dependency; not sure if it _requires_ an APE...

**Comment by astrofrog** [comment]
I have a mostly written APE for this already, just need to open the PR - thanks for the reminder!

**Comment by pllim** [comment]
FWIW I am still not sold on requiring scipy as it is still a heavyweight package and if there is something exotic enough out there (raspberry pi?) that scipy does not provide wheel for, then we're doomed.

**Comment by astrofrog** [comment]
There are scipy wheels for Raspberry Pi FWIW - https://www.piwheels.org/project/scipy/ - but I understand your point.

**Comment by astrofrog** [comment]
Thinking about this more, I am struggling to find examples of platforms/OSes where installing scipy is not easy. I have sent an email to astropy-dev to find out if anyone has specific examples, because if there aren't any, maybe times have moved on and scipy is fine to install and can be added as a required dependency without having to worry.

**Comment by mhvk** [comment]
I still feel like I wrote above, https://github.com/astropy/astropy/issues/12448#issuecomment-1003152094: I dislike the idea that the installed size more than doubles; I don't really see what is wrong with just recommending `scipy` (as we currently do). 

It also seems to me one could get most of the desired simplification by simply failing on import of the submodules that need it (i.e., stop attempting to, e.g., make `cosmology` work also without scipy).

**Comment by astrofrog** [comment]
@mhvk - according to my examples above that would mean disabling:

* astropy.coordinates
* astropy.cosmology
* astropy.units
* astropy.convolution
* astropy.modeling

and any sub-packages that depend on these, if scipy is not installed!

**Comment by astrofrog** [comment]
I guess that fundamentally I see the use of optional dependency as a loophole - we don't want to require it to keep the install small but basically people will still have to install it to use a lot of things so we didn't **really** save space.

**Comment by pllim** [comment]
There are still people who install astropy just to use `astropy.io.fits`. 😉

**Comment by mhvk** [comment]
`units` is just when used in combination with  cosmology example, so it would (and should) be caught on import of `cosmology`.  `coordinates` has only the matching. I think only `cosmology` and `modeling ` really *rely* in scipy, with `convolution` a more half-way example.

**Comment by mhvk** [comment]
p.s. Really the problem is that `pip` is a really lousy package manager, where it is not possible to have a minimal install that removes some of the dependencies. Now it does seem the whole packaging is being rethought; maybe we should just wait on that.

**Comment by dhomeier** [comment]
Even `modeling` is not sold hard on scipy; > 60 % of its tests are still running with `not HAS_SCIPY`.

**Comment by taldcroft** [comment]
I don't think that failing import of entire sub-packages is useful. The current fine-grained approach of disabling specific functionality and raising an exception that clearly states that scipy needs to be installed seems reasonable and not confusing.

I think in practice the vast majority of astropy users end up having scipy installed anyway so the package size probably doesn't matter for a typical user. But could there be niche users on a severely disk-limited systems that are impacted? Generally speaking of course 85 Mb is nothing these days.  

I'm also curious about any impact on CI since that means many more downloads of the large scipy package. On the flip side, if a lot of tests require scipy and are not done in the normal test matrix (for current req'd deps) then we are potentially missing bugs.

**Comment by saimn** [comment]
The benefit of having scipy as a required dependency would also be to simplify the code, no more need to check if scipy is available, no `HAS_SCIPY`, no conditional tests. Having a way to recommend some dependency does not solve that since the import should still be optional in the code and tests.

So maybe something we could do is to make scipy required by some subpackage, e.g. `modeling`. When importing `astropy.modeling` it would raise a nice and informative error if scipy is missing, and then we could remove the checks in the submodules and tests. (And even if many `modeling` tests currently run without scipy, the subpackage itself is pretty useless without scipy).

On other hand, I guess astropy is used (mostly for `io.fits` ?) by observatories on some low level workstations and/or archive servers where admins would prefer to keep the number of dependencies small ?

**Comment by mhvk** [comment]
@saimn - indeed, I was suggesting the error on importing modules that use scipy heavily so that we can get the code simplification without having to add the requirement for all of astropy. One could use the same for `visualization` without matplotlib.

Note that at the end on the long thread, there had been convergence on how to enable indicating smaller subsets in `pip`, and the suggestion to write a PEP -- though in December the main author noted that he would have no time and preferred to be a co-author -- https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/138

**Comment by mhvk** [comment]
Related: I just got a message from someone at RIKEN who asked whether it was allowed to pull `astropy.units` out to avoid the cost of all the rest of astropy. (Of course, it is, and I'll write to say so, noting that it means giving up on bug fixes, etc.) Maybe this is an example of users already finding that too much is bundled -- right now of course we do not hear from those who have reasons for a minimal system and would be like to use just quantities or tables+io or  just the actual basic astronomy stuff, `Time` and `SkyCoord`.

Anyway, I guess there are 3 options:
1. Keep things as they are;
2. Require scipy;
3. Raise for some modules if scipy is not present.

I think in the end it really depends on the problem we're trying to solve. I don't think we are making things particularly difficult for users, since, as @taldcroft noted, most have `scipy` installed anyway. And it would seem both that, and complications inside modules, would be solved just as well by (3) as by (1) - and (3) would mean no changes needed if `pip` ever allowed one to remove dependencies from the default list.

p.s. I think the by-type-of-machine route is more likely to lead to confusion, but perhaps a fourth option is to install `scipy` automatically on `mac` and `windows`, since `linux`  users (including CI) can be assumed to need less hand-holding (maybe include `conda` for automatic install too).

**Comment by mhvk** [comment]
Note that we do have `pip install astropy[recommended]` or `pip install astropy[all]` for those who need everything. Perhaps that should just be more prominent (including in the import errors)? Similarly, before making astropy into a new iraf, we could have `pip install astropy-all` that also pulls in all the affiliated packages; absolutely nothing stopping someone from creating that!

p.s. As I wrote before, I'd be all for making `scipy` a requirement if `pip install astropy[minimal]` actually existed; we're having a very long discussion mostly because python's package system is bad. Something that will likely be fixed soon enough, since it is even worse for packages that link to compiled libraries.

**Comment by mhvk** [comment]
p.s. Splitting out `astropy.units` has come up multiple times; e.g., https://github.com/astropy/astropy/issues/10606
It certainly isn't particularly astronomy specific.

**Comment by pllim** [comment]
I agree with @mhvk here on not requiring scipy. I think there are very different groups of people with very different requirements. A typical user might want everything already there with a single pip call, but if you are deploying stuff using astropy on the cloud, storage is precious and doubling the install size would be bad.

ref: https://github.com/astropy/astropy/issues/12448#issuecomment-1601680781

**Comment by jehturner** [comment]
@astrofrog, I think the ease of installation of SciPy depends on what BLAS you want to use and whether it available from your OS or PyPI / Anaconda etc. When I used to install all these things from source a decade or two ago, I recall ATLAS being a bit of a pain to compile and also SciPy itself requiring C, ForTran **and** C++ compilers to build (the last one being an undocumented build requirement at the time). From a quick poke around, it looks like the SciPy on PyPI (someone please complete the rhyme) bundles its own copy of OpenBLAS, so installation should be straightforward as long as you're happy with that; presumably you will be if you just want AstroPy to be installable -- though perhaps users should be aware that it might not have optimized performance out of the box (I don't have up-to-date info. on that).

**Comment by jehturner** [comment]
Note that if I try to install `scipy==0.8.0` from 2010 with `pip`, the latter fails with a `BlasNotFoundError`. So I think the answer is that installing it from PyPI is easier than it used to be, because the required dependencies are bundled nowadays, rather than sought from the OS (at the expense of hiding optimization options from the user).

**Comment by dhomeier** [comment]
> p.s. Splitting out `astropy.units` has come up multiple times; e.g., #10606
> It certainly isn't particularly astronomy specific.

Quite naturally, it would make more sense to have it available as an external dependency building directly on numpy or a generalised array interface. Most recently under discussion, I think, in https://discuss.scientific-python.org/t/physical-units-summit/390

**Comment by pllim** [comment]
Careful now... there is a dedicated balance between these 3 scenarios:

1. Breaking everything back into their own separate packages. Remember that we made a conscious decision to put "core" functionality in the same package so users don't have to install a million things to get stuff working.
2. Bringing back namespace packages (gasp).
3. Recreating IRAF but in Python.
